extensions/panache/hibernate-orm-rest-data-panache/deployment/src/main/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/RestDataEntityInfoProvider.java

```diff
@@ -0,0 +1,86 @@
+package io.quarkus.hibernate.orm.rest.data.panache.deployment;
+
+import javax.persistence.Id;
+
+import org.hibernate.bytecode.enhance.spi.EnhancerConstants;
+import org.jboss.jandex.ClassInfo;
+import org.jboss.jandex.DotName;
+import org.jboss.jandex.FieldInfo;
+import org.jboss.jandex.IndexView;
+import org.jboss.jandex.MethodInfo;
+
+import io.quarkus.deployment.bean.JavaBeanUtil;
+import io.quarkus.gizmo.MethodDescriptor;
+import io.quarkus.hibernate.orm.panache.PanacheEntityBase;
+import io.quarkus.rest.data.panache.deployment.RestDataEntityInfo;
+
+final class RestDataEntityInfoProvider {
+
+    private static final DotName PANACHE_ENTITY_BASE = DotName.createSimple(PanacheEntityBase.class.getName());
+
+    private final IndexView index;
+
+    RestDataEntityInfoProvider(IndexView index) {
+        this.index = index;
+    }
+
+    RestDataEntityInfo get(String type) {
+        ClassInfo classInfo = index.getClassByName(DotName.createSimple(type));
+        FieldInfo idField = getIdField(classInfo);
+        return new RestDataEntityInfo(classInfo.toString(), idField, getSetter(classInfo, idField));
+    }
+
+    private FieldInfo getIdField(ClassInfo classInfo) {
+        ClassInfo tmpClassInfo = classInfo;
+        while (tmpClassInfo != null) {
+            for (FieldInfo field : tmpClassInfo.fields()) {
+                if (field.hasAnnotation(DotName.createSimple(Id.class.getName()))) {
+                    return field;
+                }
+            }
+            if (classInfo.superName() != null) {
+                tmpClassInfo = index.getClassByName(classInfo.superName());
+            } else {
+                tmpClassInfo = null;
+            }
+        }
+        throw new IllegalArgumentException("Couldn't find id field of " + classInfo);
+    }
+
+    private MethodDescriptor getSetter(ClassInfo entityClass, FieldInfo field) {
+        if (isPanacheEntity(entityClass)) {
+            return getPanacheEntitySetter(entityClass, field);
+        }
+        return getGenericEntitySetter(entityClass, field);
+    }
+
+    private boolean isPanacheEntity(ClassInfo entityClass) {
+        if (entityClass == null || entityClass.superName() == null) {
+            return false;
+        }
+        if (PANACHE_ENTITY_BASE.equals(entityClass.superName())) {
+            return true;
+        }
+        return isPanacheEntity(index.getClassByName(entityClass.superName()));
+    }
+
+    private MethodDescriptor getPanacheEntitySetter(ClassInfo entityClass, FieldInfo field) {
+        return MethodDescriptor.ofMethod(entityClass.toString(),
+                EnhancerConstants.PERSISTENT_FIELD_WRITER_PREFIX + field.name(), void.class, field.type().toString());
```

Comment:
This is not OK:

- you should target `"set"+capitalize(field.name)` rather than go directly via the Hibernate method.
- this won't work for generics:

```java
@Entity
public class MyEntity extends EntitySupertype<String> {}
@MappedSuperclass
public class EntitySupertype<IdType> extends PanacheEntityBase{
 @Id
 public IdType id;
}
```

Here, `field.type()` is going to return `IdType` even though in your case it's really `String`. Note that this issue may also affect your use of the id field without setter and the generic entity setter. A few tests will tell you if they're affected.

Reply:
Initially I tried this
```java
return MethodDescriptor.ofMethod(
    entityClass.toString(), 
    JavaBeanUtil.getSetterName(field.name()), 
    void.class, 
    field.type().toString());
```
Which in my test was 
```java
return MethodDescriptor.ofMethod(
    "io.quarkus.hibernate.orm.rest.data.panache.deployment.entity.Item", 
    "setId", 
    void.class, 
    "java.lang.Long");
``` 
and it was be called as a virtual method. But I ended up getting
```java
java.lang.NoSuchMethodError: 'void io.quarkus.hibernate.orm.rest.data.panache.deployment.entity.Item.setId(java.lang.Long)'
```

Reply:
This is weird, this setter should exist.

Reply:
I've changed this bit to use the ID type specified in the generic instead of `field.type()` (e.g. Long in this case `public interface ItemsController extends PanacheEntityResource<Item, Long>`). And I've added and extra abstract class to verify that the scenario you gave above with `IdType` works too.

However, I still get `NoSuchMethodException` unless I use `EnhancerConstants.PERSISTENT_FIELD_WRITER_PREFIX` so I kept it there.

Reply:
OK, so what throws? I bet it's using reflection on the non-instrumented class.

Reply:
The thing is that we allow users to add setters to their public fields with the promise that they will be used by the users of the fields.

So ideally, if you had this entity, your setter should be invoked:

```java
@Entity
public class MyEntity extends PanacheEntityBase {
 @Id
 public Long id;

 public void setId(Long id){
  System.err.println("We're called");
  this.id = id;
 }
}
```

Even when all clients call `entity.id = 23`.

Reply:
I've replaced the setter lookup logic to the following:
```java
    private MethodDescriptor getSetter(ClassInfo entityClass, FieldInfo field) {
        if (entityClass == null) {
            return null;
        }
        MethodInfo methodInfo = entityClass.method(JavaBeanUtil.getSetterName(field.name()), field.type());
        if (methodInfo != null) {
            return MethodDescriptor.of(methodInfo);
        } else if (entityClass.superName() != null) {
            return getSetter(index.getClassByName(entityClass.superName()), field);
        }
        return null;
    }
```

I tried locally with a Panache entity and a simple POJO entity. I also tried with a custom setter method provided by the user. Let's see if the CI brings some issues though.

Before I tried calling a setter with `MethodDescriptor.ofMethod(className, JavaBeanUtil.getSetterName(fieldName), void.class, fieldType);` which didn't work.